components/connector-service/internal/certificates/headerparser.go

```diff
@@ -0,0 +1,126 @@
+package certificates
+
+import (
+	"github.com/kyma-project/kyma/components/connector-service/internal/apperrors"
+	"net/http"
+	"regexp"
+)
+
+const ClientCertHeader = "X-Forwarded-Client-Cert"
+
+type HeaderParser interface {
+	ParseCertificateHeader(r http.Request) (CertInfo, apperrors.AppError)
+}
+
+type headerParser struct {
+	country      string
+	locality     string
+	organization string
+	unit         string
+	province     string
+	central      bool
+}
+
+type CertInfo struct {
+	Hash    string
+	Subject string
+}
+
+func NewHeaderParser(country, province, locality, organization, unit string, central bool) HeaderParser {
+	return &headerParser{
+		country:      country,
+		locality:     locality,
+		organization: organization,
+		unit:         unit,
+		province:     province,
+		central:      central,
+	}
+}
+
+func (hp *headerParser) ParseCertificateHeader(r http.Request) (CertInfo, apperrors.AppError) {
+	certHeader := r.Header.Get(ClientCertHeader)
+
+	if certHeader == "" {
+		return CertInfo{}, apperrors.BadRequest("Certificate header is empty")
+	}
+
+	subjectRegex := regexp.MustCompile(`Subject="(.*?)"`)
+
+	subjects := extractFromHeader(certHeader, subjectRegex)
+
+	hashRegex := regexp.MustCompile(`Hash=([0-9a-f]*)`)
+
+	hashes := extractFromHeader(certHeader, hashRegex)
+
+	certInfos := createCertInfos(subjects, hashes)
+
+	if hp.central {
+		return hp.getCertInfoWithNonEmptySubject(certInfos)
+	} else {
+		return hp.getCertInfoWithMatchingSubject(certInfos)
+	}
+}
+
+func extractFromHeader(certHeader string, regex *regexp.Regexp) []string {
+	var matchedStrings []string
+
+	matches := regex.FindAllStringSubmatch(certHeader, -1)
+
+	for _, match := range matches {
+		hash := get(match, 1)
+		matchedStrings = append(matchedStrings, hash)
+	}
+
+	return matchedStrings
+}
+
+func get(array []string, index int) string {
+	if len(array) > index {
+		return array[index]
+	}
+	return ""
+}
+
+func createCertInfos(subjects, hashes []string) []CertInfo {
+	certInfos := make([]CertInfo, 0)
+	for i := 0; i < len(subjects); i++ {
+		certInfo := newCertInfo(subjects[i], hashes[i])
+		certInfos = append(certInfos, certInfo)
+	}
+	return certInfos
+}
+
+func (hp *headerParser) getCertInfoWithNonEmptySubject(infos []CertInfo) (CertInfo, apperrors.AppError) {
```

Comment:
Those two functions: `getCertInfoWithNonEmptySubject` and `getCertInfoWithMatchingSubject` could be marged to one if we refactor subject verification as stated in one of the above comments.